crates/aptos-logger/src/speculative_log.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate as aptos_logger;
+/// Infrastucture for logging speculatively, i.e. ability to clear prior logs.
+use crate::{debug, error, info, trace, warn, Level};
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_log_derive::Schema;
+use aptos_state_view::StateViewId;
+use aptos_types::transaction::Version;
+use arc_swap::ArcSwap;
+use crossbeam::utils::CachePadded;
+use once_cell::sync::Lazy;
+use rayon::prelude::*;
+use serde::Serialize;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+#[derive(Default)]
+struct BufferedLog {
+    entries: Vec<(Level, AdapterLogSchema, String)>, // Log level & the message string
+    // critical error count associated with speculative (but not cleared) logs.
+    critical_err_cnt: usize,
+    // To avoid de-allocations on the critical path, store cleared number.
+    skip_cleared: usize,
+}
+
+type LOGS = Vec<CachePadded<Mutex<BufferedLog>>>;
+
+// .1 (AtomicUsize type) tracks the total number of speculative critical errors.
+// This is done so flush_speculative_logs method can return the number of
+// critical errors immediately without traversing over all locks - that is happened
+// off the critical path in order to flush the actual messages.
+static BUFFERED_LOGS: Lazy<ArcSwap<(LOGS, AtomicUsize)>> =
+    Lazy::new(|| ArcSwap::new(Arc::new((LOGS::new(), AtomicUsize::new(0)))));
+
+#[derive(Schema, Clone)]
+pub struct AdapterLogSchema {
```

Comment:
I would have expected this context or schema to be generic, given that this lives in the `aptos-logger`. But, this feels very specific to the VM code? Perhaps a trait that could be instantiated by each caller would be cleaner? i.e., why does the `aptos-logger` need to know about block ids, first version, etc?